File: cashplot/totals.py
from cashplot.util import *
import datetime
import calendar
from decimal import Decimal
from enum import Enum


class Period(Enum):
    YEAR = 1
    QUARTER = 2
    MONTH = 3
    WEEK = 4
    DAY = 5


def get_categories(tr_balances):
    """Return all categories occurring in the transactions+balances."""
    categories = set(map(lambda tr: tr.category, tr_balances))
    for account_name in tr_balances[0].balances.keys():
        if account_name in categories:
            categories.remove(account_name)
    return sorted(list(categories))


def floor_period(date, period):
    """Floor the given date down to the first of the period."""
    if period == Period.YEAR:
        return datetime.datetime(date.year, 1, 1).date()
    elif period == Period.QUARTER:
        return datetime.datetime(date.year, date.month - (date.month - 1) % 3, 1).date()
    elif period == Period.MONTH:
        return datetime.datetime(date.year, date.month, 1).date()
    elif period == Period.WEEK:
        year, week, day = date.isocalendar()
        return datetime.date.fromisocalendar(year, week, 1)
    elif period == Period.DAY:
        return date
    else:
        raise ValueError('unknown period type')


def next_period(date, period):
    """Given a date, return the first day of the next period."""
    # Make sure the date is the first of the current period.
    date = floor_period(date, period)
    if period == Period.YEAR:
        new_year = date.year + 1
        new_month = 1
        new_day = 1
    elif period == Period.QUARTER:
        new_year = date.year
        new_month = date.month + 3
        new_day = 1
    elif period == Period.MONTH:
        new_year = date.year
        new_month = date.month + 1
        new_day = 1
    elif period == Period.WEEK:
        new_year = date.year
        new_month = date.month
        new_day = date.day + 7
    elif period == Period.DAY:
        new_year = date.year
        new_month = date.month
        new_day = date.day + 1
    else:
        raise ValueError('unknown period type')

    if new_month > 12:
        new_year += 1
        new_month -= 12
    if new_day > calendar.monthrange(new_year, new_month)[1]:
        new_day -= calendar.monthrange(new_year, new_month)[1]
        new_month += 1
    if new_month > 12:
        new_year += 1
        new_month -= 12

    return datetime.datetime(new_year, new_month, new_day).date()


def period_thirds(date, period):
    """
    Given a date, return the dates of 1/3rds and 2/3rds through the period.

    NOTE: If period == Period.DAY, returns a datetime instead of a date
    """
    # Make sure the date is the first of the current period.
    date = floor_period(date, period)
    next_date = next_period(date, period)
    if period == Period.DAY:
        date = datetime.datetime(date.year, date.month, date.day)
        next_date = datetime.datetime(
            next_date.year, next_date.month, next_date.day)
    difference = next_date - date

    one_thirds = date + (difference / 3) * 1
    two_thirds = date + (difference / 3) * 2
    return (one_thirds, two_thirds)


def period_halves(date, period):
    """
    Given a date, return the date but half through the period.

    NOTE: If period == Period.DAY, returns a datetime instead of a date
    """
    # Make sure the date is the first of the current period.
    date = floor_period(date, period)
    next_date = next_period(date, period)
    if period == Period.DAY:
        date = datetime.datetime(date.year, date.month, date.day)
        next_date = datetime.datetime(
            next_date.year, next_date.month, next_date.day)
    difference = next_date - date

    return date + difference / 2


def override_thirds(dates, period):
    """Given the dates, return two lists of the dates but 1/3rds and 2/3rds through the period."""
    thirds = list(map(lambda date: period_thirds(date, period), dates))
    one_thirds = list(map(lambda thirds: thirds[0], thirds))
    two_thirds = list(map(lambda thirds: thirds[1], thirds))
    return one_thirds, two_thirds


def override_halves(dates, period):
    """Given the dates, return a lists of the dates but half through the period."""
    return list(map(lambda date: period_halves(date, period), dates))


def categories_changes(tr_balances, period):
    """Calculate the period total income/expenses for each category occurring in the given transactions+balances."""
    categories = get_categories(tr_balances)

    first_period = floor_period(tr_balances[0].date, period)
    last_period = floor_period(tr_balances[-1].date, period)

    income_changes = fill_dict(categories, [])
    expenses_changes = fill_dict(categories, [])

    periods = []
    cur_period = first_period
    while cur_period <= last_period:
        cur_income_changes = fill_dict(categories, Decimal(0))
        cur_expenses_changes = fill_dict(categories, Decimal(0))
        for tr in tr_balances:
            if floor_period(tr.date, period) == cur_period and tr.category in categories:
                if tr.change >= 0:
                    cur_income_changes[tr.category] += tr.change
                else:
                    cur_expenses_changes[tr.category] += -tr.change
        for category in categories:
            income_changes[category].append(cur_income_changes[category])
            expenses_changes[category].append(cur_expenses_changes[category])
        periods.append(cur_period)
        cur_period = next_period(cur_period, period)

    return (periods, income_changes, expenses_changes)


def last_year_range(tr_balances):
    """Calculate the date range from a year prior to the transaction, to the transaction"""
    last_date = tr_balances[-1].date
    last_year_date = datetime.datetime(
        last_date.year - 1, last_date.month, last_date.day).date()
    return [last_year_date, last_date]

File: src/js/lib/totals.js
/** @module */
const transactions = require("./transactions");
const utils = require("./utils");
const assert = require("nanoassert");

/** Enum denoting a span of time (YEAR/QUARTER/...). */
exports.Period = {
  YEAR: 1,
  QUARTER: 2,
  MONTH: 3,
  WEEK: 4,
  DAY: 5,
};

/**
 * @param {Array<TransactionBalance>} trBalances - Transactions+balances.
 * @return {Array<string>} All categories occurring in the transactions+balances.
 */
exports.getCategories = function(trBalances) {
  const accountNames = Object.keys(trBalances[0].balances);
  const isNotAnAccount = (category) => accountNames.indexOf(category) === -1;

  const rawCategories = trBalances.map((tr) => tr.transaction.category);
  // Remove duplicates and account names from categories.
  return [...new Set(rawCategories)].filter(isNotAnAccount);
}

/**
 * Calculate the date range from a year prior to the last transaction, to the last transaction.
 * @param {Array<TransactionBalance>} trBalances - Transactions+balances.
 * @return {Array.<Date, Date>} The dates.
 */
exports.lastYearRange = function(trBalances) {
  assert(Array.isArray(trBalances));
  assert(trBalances.length > 0);
  assert(
    trBalances.every((tr) => tr instanceof transactions.TransactionBalance)
  );
  const untilDate = trBalances[trBalances.length - 1].transaction.date;
  const afterDate = new Date(untilDate.valueOf());
  afterDate.setFullYear(afterDate.getFullYear() - 1);
  return [afterDate, untilDate];
}

/**
 * Floors the date to the first day in the given period.
 *
 * For example, given July 23rd and exports.Period.MONTH, floors it down to July 1st.
 *
 * @param {Date} date - The date.
 * @param {exports.Period} period - The period.
 * @return {Date} The floored date.
 */
exports.floorPeriod = function(date, period) {
  assert(date instanceof Date);
  switch (period) {
    case exports.Period.YEAR: {
      return new Date(Date.UTC(date.getUTCFullYear(), 0, 1));
    }
    case exports.Period.QUARTER: {
      return new Date(
        Date.UTC(
          date.getUTCFullYear(),
          date.getUTCMonth() - (date.getUTCMonth() % 3),
          1
        )
      );
    }
    case exports.Period.MONTH: {
      return new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), 1));
    }
    case exports.Period.WEEK: {
      // https://stackoverflow.com/a/4156516
      const d = new Date(date);
      const day = d.getDay();
      const diff = d.getDate() - day + (day == 0 ? -6 : 1); // Adjust when day is sunday.
      return new Date(d.setDate(diff));
    }
    case exports.Period.DAY: {
      return new Date(
        Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate())
      );
    }
    default: {
      throw new Error("Unknown period " + period);
    }
  }
}

/**
 * Given a date, return the first day of the next period.
 * @param {Date} date - The date.
 * @param {exports.Period} period - The period.
 * @return {Date} The first day in the next period.
 */
exports.nextPeriod = function(date, period) {
  assert(date instanceof Date);
  const floorDate = exports.floorPeriod(date, period);
  switch (period) {
    case exports.Period.YEAR: {
      return new Date(
        Date.UTC(
          floorDate.getUTCFullYear() + 1,
          floorDate.getUTCMonth(),
          floorDate.getUTCDate()
        )
      );
    }
    case exports.Period.QUARTER: {
      return new Date(
        Date.UTC(
          floorDate.getUTCFullYear(),
          floorDate.getUTCMonth() + 3,
          floorDate.getUTCDate()
        )
      );
    }
    case exports.Period.MONTH: {
      return new Date(
        Date.UTC(
          floorDate.getUTCFullYear(),
          floorDate.getUTCMonth() + 1,
          floorDate.getUTCDate()
        )
      );
    }
    case exports.Period.WEEK: {
      return new Date(
        Date.UTC(
          floorDate.getUTCFullYear(),
          floorDate.getUTCMonth(),
          floorDate.getUTCDate() + 7
        )
      );
    }
    case exports.Period.DAY: {
      return new Date(
        Date.UTC(
          floorDate.getUTCFullYear(),
          floorDate.getUTCMonth(),
          floorDate.getUTCDate() + 1
        )
      );
    }
    default: {
      throw new Error("Unknown period " + period);
    }
  }
}

/**
 * Given a date, return the dates of 1/3rd and 2/3rds through the period.
 * @param {Date} date - The date.
 * @param {exports.Period} period - The period.
 * @return {Array<Date>} Dates of 1/3rd and 2/3rds through the period.
 */
exports.periodThirds = function(date, period) {
  const fromDate = exports.floorPeriod(date, period);
  const untilDate = exports.nextPeriod(date, period);

  const oneThirdDate = new Date(
    fromDate.getTime() + (untilDate.getTime() - fromDate.getTime()) / 3
  );
  const twoThirdsDate = new Date(
    fromDate.getTime() + ((untilDate.getTime() - fromDate.getTime()) / 3) * 2
  );

  if (period == exports.Period.DAY) {
    return [oneThirdDate, twoThirdsDate];
  } else {
    return [
      new Date(
        Date.UTC(
          oneThirdDate.getUTCFullYear(),
          oneThirdDate.getUTCMonth(),
          oneThirdDate.getUTCDate()
        )
      ),
      new Date(
        Date.UTC(
          twoThirdsDate.getUTCFullYear(),
          twoThirdsDate.getUTCMonth(),
          twoThirdsDate.getUTCDate()
        )
      ),
    ];
  }
}

/**
 * Given a date, return the date but half through the period.
 * @param {Date} date - The date.
 * @param {exports.Period} period - The period.
 * @return {Array<Date>} Dates half through the period.
 */
exports.periodHalves = function(date, period) {
  const fromDate = exports.floorPeriod(date, period);
  const untilDate = exports.nextPeriod(date, period);

  const halfDate = new Date(
    fromDate.getTime() + (untilDate.getTime() - fromDate.getTime()) / 2
  );

  if (period == exports.Period.DAY) {
    return halfDate;
  } else {
    return new Date(
      Date.UTC(
        halfDate.getUTCFullYear(),
        halfDate.getUTCMonth(),
        halfDate.getUTCDate()
      )
    );
  }
}

/**
 * Calculate the period total income/expenses for each category occurring in the given transactions+balances.
 * @param {Array<TransactionBalance>} trBalances - The transactions+balances.
 * @param {exports.Period} period - The period by which to group income/expenses by.
 * @return {Array.<Array<Date>,Object<string,number>,Object<string,number>>} -
 *   An array of periods, an array of income totals and an array of expenses totals.
 */
exports.categoriesChanges = function(trBalances, period) {
  const categories = exports.getCategories(trBalances);

  const firstPeriod = exports.floorPeriod(trBalances[0].transaction.date, period);
  const lastPeriod = exports.floorPeriod(
    trBalances[trBalances.length - 1].transaction.date,
    period
  );

  const incomeChanges = utils.fillObject(categories, [], true);
  const expensesChanges = utils.fillObject(categories, [], true);

  const periods = [];
  let curPeriod = firstPeriod;

  while (curPeriod <= lastPeriod) {
    curIncomeChanges = utils.fillObject(categories, 0);
    curExpensesChanges = utils.fillObject(categories, 0);
    for (const trBalance of trBalances) {
      const tr = trBalance.transaction;
      if (
        exports.floorPeriod(tr.date, period).getTime() === curPeriod.getTime() &&
        categories.indexOf(tr.category) !== -1
      ) {
        if (tr.change >= 0) {
          curIncomeChanges[tr.category] += tr.change;
        } else {
          curExpensesChanges[tr.category] += -tr.change;
        }
      }
    }
    for (const category of categories) {
      incomeChanges[category].push(curIncomeChanges[category]);
      expensesChanges[category].push(curExpensesChanges[category]);
    }
    periods.push(curPeriod);
    curPeriod = exports.nextPeriod(curPeriod, period);
  }

  return [periods, incomeChanges, expensesChanges];
}
